libtransact/src/workload/batch_gen.rs

```diff
@@ -0,0 +1,398 @@
+    let pk = match signer.public_key() {
+        Ok(pk) => pk,
+        Err(err) => return Err(BatchingError::SigningError(err)),
+    };
+    let public_key = pk.as_hex();
+
+    let txn_ids = txns
+        .iter()
+        .map(|txn| String::from(txn.get_header_signature()))
+        .collect();
+    batch_header.set_transaction_ids(protobuf::RepeatedField::from_vec(txn_ids));
+    batch_header.set_signer_public_key(public_key);
+
+    let header_bytes = batch_header.write_to_bytes()?;
+    let signature = signer
+        .sign(&header_bytes)
+        .map_err(BatchingError::SigningError);
+    match signature {
+        Ok(signature) => {
+            let mut batch = Batch::new();
+            batch.set_header_signature(signature.as_hex());
+            batch.set_header(header_bytes);
+            batch.set_transactions(protobuf::RepeatedField::from_vec(txns));
+
+            Ok(batch)
+        }
+        Err(err) => Err(err),
+    }
+}
+
+pub struct SignedBatchIterator<'a> {
+    transaction_iterator: &'a mut dyn Iterator<Item = Transaction>,
+    max_batch_size: usize,
+    signer: &'a dyn Signer,
+}
+
+impl<'a> SignedBatchIterator<'a> {
+    pub fn new(
+        iterator: &'a mut dyn Iterator<Item = Transaction>,
+        max_batch_size: usize,
+        signer: &'a dyn Signer,
+    ) -> Self {
+        SignedBatchIterator {
+            transaction_iterator: iterator,
+            max_batch_size,
+            signer,
+        }
+    }
+}
+
+impl<'a> Iterator for SignedBatchIterator<'a> {
+    type Item = BatchResult;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let txns = self
+            .transaction_iterator
+            .take(self.max_batch_size)
+            .collect();
+
+        Some(batch_transactions(txns, self.signer))
+    }
+}
+
+type BatchSource<'a> = LengthDelimitedMessageSource<'a, Batch>;
+
+/// Produces get batches from length-delimited source of Batches.
+pub struct BatchListFeeder<'a> {
+    batch_source: BatchSource<'a>,
+}
+
+/// Resulting BatchList or error.
+pub type BatchListResult = Result<BatchPair, BatchReadingError>;
+
+impl<'a> BatchListFeeder<'a> {
+    /// Creates a new `BatchListFeeder` with a given Batch source
+    pub fn new(source: &'a mut dyn Read) -> Self {
+        let batch_source = LengthDelimitedMessageSource::new(source);
+        BatchListFeeder { batch_source }
+    }
+}
+
+impl<'a> Iterator for BatchListFeeder<'a> {
+    type Item = BatchListResult;
+
+    /// Gets the next Batch.
+    /// `Ok(None)` indicates that the underlying source has been consumed.
+    fn next(&mut self) -> Option<Self::Item> {
+        let batches = match self.batch_source.next(1) {
+            Ok(batches) => batches,
+            Err(err) => return Some(Err(BatchReadingError::MessageError(err))),
+        };
+
+        let batch_proto = match batches.get(0) {
+            Some(batch_proto) => batch_proto,
+            None => return None,
+        };
+
+        // we have already checked that there is
```

Comment:
Could you clarify this comment?

Reply:
That comment should actually just be removed